### PR TITLE
KCM: Document KCM debugging

### PR DESCRIPTION
**Pull request by aplopez**
Added some KCM-specific details to the already present debug information.

### CODE REVIEW

File: src/contrib/debugging-sssd.rst [hunk 1]
```diff
@@ -206,3 +211,80 @@ i.e. ``su admin@ipa.test``. Then start gdb and attach it to the server:
     debugging the child once it is forked off the parent. See `gdb manual
     <https://sourceware.org/gdb/onlinedocs/gdb/Forks.html>`__ for more
     information.
+
+Debugging KCM
+*************
+KCM is different from other SSSD processes in the way it is launched. It is not
+forked by the monitor as all the other processes, but it is launched by a
+``systemd trigger unit`` when a client application (like ``klist``) opens its
+socket. After 5 minutes of inactivity it will automatically shut down.
```
Comment (justin-stephenson) [inline src/contrib/debugging-sssd.rst:220]: if `tgt_renewal = true` then we disable the responder idle timeout, maybe not worth mentioning though

File: src/contrib/debugging-sssd.rst [hunk 2]
```diff
@@ -206,3 +211,80 @@ i.e. ``su admin@ipa.test``. Then start gdb and attach it to the server:
...
+
+Once it is active (running) it is possible to attach a debugger as described
+above, but it is also possible to launch it manually (instead of waiting for
+``systemd`` to trigger it). A ``systemd trigger`` is a special type of unit that
+will start a ``systemd service`` when a certain event happens. For KCM the event
+is a client opening the socket. The service is ``sssd-kcm.service`` and is
+triggered by ``sssd-kcm.socket``.
+
+To manually launch KCM, you just need to start the service:
```
Comment (justin-stephenson) [inline src/contrib/debugging-sssd.rst:229]: s/you just need to//

File: src/contrib/debugging-sssd.rst [hunk 3]
```diff
@@ -206,3 +211,80 @@ i.e. ``su admin@ipa.test``. Then start gdb and attach it to the server:
...
+``client_idle_timeout`` option should be set to a large number of seconds.
+This value specifies how long a client can sit idle before being disconnected.
+
+.. code-block:: ini
+
+    [kcm]
+    responder_idle_timeout=0
+    client_idle_timeout=10000
+
+KCM can also be launched from the command line. For that, you first need to
```
Comment (justin-stephenson) [inline src/contrib/debugging-sssd.rst:248]: s/you first need to//

File: src/contrib/debugging-sssd.rst [hunk 4]
```diff
@@ -206,3 +211,80 @@ i.e. ``su admin@ipa.test``. Then start gdb and attach it to the server:
...
+disable the ``systemd trigger`` and stop any active instance of KCM.
+
+.. code-block:: console
+
+    $ sudo systemctl stop sssd-kcm.socket
+    $ sudo systemctl stop sssd-kcm.service
+
+If you need to watch KCM's start up, you can launch it from the command line
```
Comment (justin-stephenson) [inline src/contrib/debugging-sssd.rst:256]: s/you can//

File: src/contrib/debugging-sssd.rst [hunk 5]
```diff
@@ -206,3 +211,80 @@ i.e. ``su admin@ipa.test``. Then start gdb and attach it to the server:
...
+with some useful debug options:
+
+.. code-block:: console
+
+    $ sudo /usr/libexec/sssd/sssd_kcm --uid 0 --gid 0 --logger=stderr --debug-level=9
+
+To debug KCM from its start, you can simple launch ``sssd_kcm`` in the debugger:
```
Comment (justin-stephenson) [inline src/contrib/debugging-sssd.rst:263]: s/you can simple//

**Comment by justin-stephenson** [comment]
I added some inline comments but changes look good to me overall.

A couple other things I found useful when debugging KCM issues. You can determine if they would be worthwhile to add here or not.

- KCM opcodes and parameters from https://github.com/krb5/krb5/blob/master/src/include/kcm.h
- Strace command to dump of the KCM socket data (protocol request/response), intermixed with sssd KCM debug logging:

~~~
sudo strace -p $(pidof sssd_kcm) -Ttxvs 1024 -e read=$fds -e read,readv,write
~~~

File: src/contrib/debugging-sssd.rst (hunk 2 above)
Comment (aplopez) [reply src/contrib/debugging-sssd.rst:229]: Done

File: src/contrib/debugging-sssd.rst (hunk 3 above)
Comment (aplopez) [reply src/contrib/debugging-sssd.rst:248]: Done

File: src/contrib/debugging-sssd.rst (hunk 4 above)
Comment (aplopez) [reply src/contrib/debugging-sssd.rst:256]: Done

File: src/contrib/debugging-sssd.rst (hunk 5 above)
Comment (aplopez) [reply src/contrib/debugging-sssd.rst:263]: Done

File: src/contrib/debugging-sssd.rst [hunk 6]
```diff
@@ -206,3 +211,80 @@ i.e. ``su admin@ipa.test``. Then start gdb and attach it to the server:
...
+
+.. code-block:: console
+
+    $ sudo gdb /usr/libexec/sssd/sssd_kcm
+    (gdb) break main
+    (gdb) run --uid 0 --gid 0 --logger=files
+
+Alternatively, you can launch ``sssd_kcm`` with ``strace``:
```
[the remark this answers is not in the document]
Comment (aplopez) [reply src/contrib/debugging-sssd.rst:271]: Done

**Comment by aplopez** [comment]
>     * KCM opcodes and parameters from https://github.com/krb5/krb5/blob/master/src/include/kcm.h

I didn't add this one as it is easily found and the text mainly focuses on debugging techniques.

> ```
> sudo strace -p $(pidof sssd_kcm) -Ttxvs 1024 -e read=$fds -e read,readv,write
> ```
But I did add this one.

Thanks.

**Review by justin-stephenson — APPROVED** [review]
Ack, thank you.

**Pull request merged**